substrate/client/rpc/src/state/state_full.rs

```diff
@@ -217,6 +237,20 @@ where
 			.map_err(client_err)
 	}
 
+	fn storage_diff(
+		&self,
+		block : Block::Hash,
+		prefixes: Option<Vec<StorageKey>>
+	) -> std::result::Result<(StorageCollection, ChildStorageCollection), Error> {
+		let (mut storage_collection, mut child_storage_collection) = self.client.storage_updates_at(block).map_err(client_err)?;
+
+		// retain only required prefixes
+		storage_collection.retain(|key| self.is_target_key(sc_client_api::StorageKey(key.0.clone()), prefixes.clone()));
+		child_storage_collection.retain(|key| self.is_target_key(sc_client_api::StorageKey(key.0.clone()), prefixes.clone()));
```

Comment:
This looks like a not correct filtering of keys by prefixes.

Example
- we have a storage collection 
- Collection has 2 keys (address, address) and stores some struct
- in the past block an element of that collection was changed.
- we will see in the storage that the storage element with the key KEY1_KEY2_KEY3 has changed. 
  - the full key is  a result of concatenation of 3 keys
  - KEY1 - corresponds to the collection
  - KEY2 and KEY3 - correspond to the keys (address, address) of a particular storage element 

Filtering by the prefixes should be done this way:
- we build a full key KEY1_KEY2_KEY3 by concatenating the key of a collection with the key(s) of storage element
- then we test the full key against provided list of prefixes

In the current solution you make 2 tests:
- test KEY1 against prefixes
- test KEY2_KEY3 against prefixes
This will yield wrong results.

Please read this article before going further:
https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/